Selector API 리뷰 PR입니다.

seyoon-kim opened the pull request:
**데모페이지**
[데모페이지 이동](http://10.78.65.35:8000/examples/)

**소개**
- Selector API의 구현인 querySelector와 querySelectorAll를 직접 구현한다.
- 실제 구현 전에 요구사항에 맞는 TC를 우선 작성 후 공유한다.

**요구사항**
- getElementById, getElementByTagName을 이용해 구현한다. (특정 IE에서 이슈가 있는 getElementByClassName도 직접 구현한다.)
- .class, #id, tag 3개의 셀렉터만 이용할 수 있다.
- 셀렉터의 조합으로 엘리먼트를 찾을 수 있다 배열로 엘리먼트를 직접 반환함.

```
domutil.querySelectorAll('.class');
domutil.querySelector('#id');

//myClass라는 클래스를 가진 엘리먼트의 자식요소중 div를 모두 리턴
domutil.querySelectorAll('.myClass div');

```

seyoon-kim committed:
환경설정 파일 변경 및 소스코드 업로드

seyoon-kim committed:
Update, Demo page작성 및 소스정리

- Demo page작성
- 소스코드 정리

seyoon-kim committed:
Update, _findElementsOfmatchingSelector 함수 수정

- 셀렉터가 ID인 경우 getElementById()함수를 사용해야 합니다. 그런데 앞에 document가 아니라 rootEle라는 전에 검색된 요소의 배열로 요청되어 함수가 실행이 안되었습니다. 그래서 rootEle -> document 수정하였습니다.

seyoon-kim committed:
Update, 셀렉터가 id인 경우 해당하는 요소를 찾지못한 경우

셀렉터가 id인 경우에 해당하는 요소를 찾지 못한 경우 빈배열을 반환하도록 수정하였습니다.

seyoon-kim committed:
Update, 리뷰사항 수정

1. pakage.json 로컬웹서버 글로벌설치로 인한 local-web-sever 삭제
2. spec.js _getElementsByClassNamePolyfill 테스트케이스 안에 함수 생성하여 테스트
3. rClasName, rClassName 오타수정
4. result 변수 배열로 초기화

seyoon-kim committed:
Update, querySelectorAll의 반환값 배열 요소중 중복되는 요소 제거

querySelectorAll의 반환값의 배열 요소가 중복도는 요소가 발견되었습니다. 이를 해결 하기 위해 removeSameElement라는 함수를 만들어 중복되는 값을 제거하도록 하였습니다.

seonim-ryu commented:
![2017-09-14 11 30 36](https://user-images.githubusercontent.com/18183560/30409475-32ad315e-9940-11e7-819d-9f00535e0762.png)

아이디 셀렉터(`#`)일 때는 배열값이 반환되지 않도록 처리가 필요하겠네요~

seonim-ryu commented on src/index.js:
```diff
@@ -0,0 +1,159 @@
+
+        i = 0;
+        allElementsLength = allElements.length;
+        for (i; i < allElementsLength; i += 1) {
```
반복문 안에 `i;` 이 부분은 잘못된 사용이네요~

seyoon-kim replied:
for (; i < allElementsLength; i += 1)  이런식으로 i를 지워야 하는건가요?

youjung-hong replied:
여기 참고하시면 되겠네용!

* https://developer.mozilla.org/ko/docs/Web/JavaScript/Guide/Loops_and_iteration
* https://opentutorials.org/module/570/4962

kyuwoo-choi commented on src/index.js:
```diff
@@ -0,0 +1,159 @@
+var Domutil = (function() {
+    // getElementsByClassName*() IE 8 이하 버전 호환
+    var _getElementsByClassNamePolyfill = function(rootEle, selector) {
+        var allElements;
```
* 변수 선언과 할당은 순서 바꿔주는것이 좋겠네요.
```
var allElements;
var rClassName;
...
var result = [];
```

* 선택이지만 선언만 하는 경우는 한줄로 해도, 지금처럼 풀어도 무방합니다.

kyuwoo-choi replied:
`rClassName`: 동식선임님이 헝가리안 노테이션 얘기 꺼내신 이유가 여기있었군요.
우리가 가지고 있는건 `$` jquery하나만 지키고 있죠?
전 regex나 다른 타입들 필요한지는 약간 의문입니다...
다른분들 의견 어떤지도 궁금하네요.

seyoon-kim replied:
변수가 초기화된것을 먼저 선언하도록 하겠습니다.

dongsik-yoo commented on src/index.js:
```diff
@@ -0,0 +1,159 @@
+    // getElementsByClassName*() IE 8 이하 버전 호환
+    var _getElementsByClassNamePolyfill = function(rootEle, selector) {
+        var allElements;
+        var result = [];
```
변수 초기화 코드가 여기만 있네요. 통일성 있게 가는 게 좋을 것 같습니다.

seyoon-kim replied:
넵 확인하도록 하겠습니다

seonim-ryu commented on src/index.js:
```diff
@@ -0,0 +1,159 @@
+    var _getElementsByClassNamePolyfill = function(rootEle, selector) {
+        var allElements;
+        var result = [];
+        var rClassName;
```
`rClassName` 보다 좀 더 명확한 변수명을 사용하는게 좋습니다~

seyoon-kim replied:
저는 className을 찾는 정규식이다 라고 생각해서 rClassName으로 생각했는데요 좀 더 명확 하다면 rFindClassName이런식으로 하면 좋을까요?

seonim-ryu replied:
변수나 함수명은 길어도 좋구요. `regex` 이렇게 풀어서 쓰는게 좋을듯 합니다ㅎ

seyoon-kim replied:
네 명확하게 작성하도록 하겠습니다

seonim-ryu commented on src/index.js:
```diff
@@ -0,0 +1,159 @@
+    /*
+     * 1. selector의 앞에 문자열에 따라 id, class, tagName을 구별하여 해당하는 엘리멘트들의 배열을 반환하는 함수
+     * 2. rootEle, selector 두 개의 인자를 받는다. rootEle는 이전에 받은 결과값으로서 해당 값으로 다시 다음번째의 selector 값을 이용하여 값을 찾을때 사용한다.
+     */
```
주석도 컨벤션에 맞게 변경해야겠네요~
eslint룰에 걸리지 않던가요 ;ㅅ;

seyoon-kim replied:
제가 eslint에서 확인할 때는 이상이 없었습니다. eslint 설정을 다시 확인해보도록 하겠습니다

youjung-hong replied:
함수를 값으로 선언해서 eslint가 걸러주지 않은 것 같네요~
함수 위에 `jsdoc` 형태로 주석을 작성하는 데요.
그 규칙에 따라 작성하면 API 문서가 자동으로 만들어집니다.

* https://eslint.org/docs/rules/require-jsdoc - 
* https://eslint.org/docs/rules/valid-jsdoc

seyoon-kim replied:
올려주신 문서대로 주석 처리 하도록 하겠습니다.

seonim-ryu commented on src/index.js:
```diff
@@ -0,0 +1,159 @@
+    var _findElementsOfMatchingSelector = function(rootEle, selectors) {
+        var result = [];
+        var rClassName = /^\./g;
+        var rIdName = /^#/g;
```
이런 정규식 패턴들은 함수 내에서 매번 할당해서 사용하는 것보다 상수 값으로 빼서 관리하는게 좋습니다~

seyoon-kim replied:
넵 수정하도록 하겠습니다

minkyu-yi commented on src/index.js:
```diff
@@ -0,0 +1,159 @@
+        i = 0;
+        allElementsLength = allElements.length;
+        for (i; i < allElementsLength; i += 1) {
+            if (rClassName.test(allElements[i].className)) {
```
`my-list`라는 클래스명을 갖는 엘리먼트랑 `list`라는 클래스명을 갖는 엘리먼트가 있을 때
getElementsByClassName('list')를 수행하면 `my-list`도 같이 찾아질것 같아요

seyoon-kim replied:
넵 해당 클래스 테스트해서 확인 하도록 하겠습니다

minkyu-yi commented on src/index.js:
```diff
@@ -0,0 +1,159 @@
+var Domutil = (function() {
+    // getElementsByClassName*() IE 8 이하 버전 호환
+    var _getElementsByClassNamePolyfill = function(rootEle, selector) {
```
selector보단 그냥 name이나  className으로 표현하는게 더 명확할 것 같아요.
`list`라는 클래스명이 있을 때
selector는 `.list`라고 생각하기 쉬운데,
실제로 getElementsByClassName에서 인자로 받는것은 `list`라서 혼동이 있을 것 같아요

seyoon-kim replied:
넵 알겠습니다.

dongsik-yoo commented on test/spec.js:
```diff
@@ -0,0 +1,129 @@
+
+    // 셀렉터가 class인 경우, 해당 class를 가지고 있는 엘리멘트 배열을 반환
+    it('if selector is class, should be return array class element', function() {
+        document.body.innerHTML = '<div class="main">main</div><div>no main</div>';
```
테스트에서 반복으로 환경이 설정되어야 할 경우 `beforeEach`를 쓰거나 `fixture`를 사용하는 편이 낫습니다.

seyoon-kim replied:
네 앞으로 말씀해주신 반복적인 부분들은 beforeEach나 fixture를 사용도록 하겠습니다.

seonim-ryu commented on src/index.js:
```diff
@@ -0,0 +1,159 @@
+        var tempLength;
+
+        x = 0;
+        fromLength = from.length;
```
`x`, `fromLength`는 변수 선언과 동시에 초기화해도 되겠네요ㅎ
그리고 변수 선언만 하는 경우 `var`로 하나씩 선언하지 않아도 됩니다~
`var y, count, tempLength;` 이렇게 한 줄에 선언해도 되요~

seyoon-kim replied:
넵 알겠습니다

seonim-ryu commented on src/index.js:
```diff
@@ -0,0 +1,159 @@
+
+        x = 0;
+        fromLength = from.length;
+        for (x; x < fromLength; x += 1) {
```
여기도 `x;` 사용이 잘못되었네요~

seyoon-kim replied:
네 앞서 말씀해주신 사항으로 고치도록 하겠습니다

dongsik-yoo commented on test/spec.js:
```diff
@@ -0,0 +1,129 @@
+});
+
+describe('getElementsByClassName Pollyfil', function() {
+    var _getElementsByClassNamePolyfill = function(rootEle, selector) {
```
내부에 선언되어 있는 내부용 함수를 테스트 하기 위해서는 함수를 외부로 노출 시켜서 테스트 하셔도 됩니다. 그 편이 코드 중복으로 들어가는 것보다는 낫습니다.

seyoon-kim replied:
넵 테스트할 코드는 외부로 노출하여 하도록 하겠습니다

seyoon-kim replied:
넵 테스트할 코드는 외부로 노출 시켜서 사용하도록 하겠습니다.

seonim-ryu commented on src/index.js:
```diff
@@ -0,0 +1,159 @@
+                y = 0;
+                tempLength = temp.length;
+                for (y; y < tempLength; y += 1) {
+                    if (temp[y] === from[x]) {
```
`x`, `y` 대신 다른 변수명으로 변경하는게 좋을 것 같습니다~
코드가 잘 안읽혀서요 ;ㅅ;

seyoon-kim replied:
넵 어떤 루프의 변수인지 알수 있도록 fromNum, tempNum 이런식으로 바꾸도록 하겠습니다.

youjung-hong commented on src/index.js:
```diff
@@ -0,0 +1,159 @@
+        var result = [];
+        var rClassName;
+        var i;
+        var allElementsLength;
```
초기화되어있는 값을 먼저 선언하는 것이 우리 팀 규칙입니다.

```javascript
var result = [];
var allElements;
var rClassName;
var i;
var allElementsLength;
```

seyoon-kim replied:
넵 말씀해주신 부분 앞으로 적용하도록 하겠습니다

kyuwoo-choi commented on src/index.js:
```diff
@@ -0,0 +1,159 @@
+
+        founded = [];
+        from = [document];
+        arrSeletor = selectors.split(' ');
```
스페이스 한칸으로 split하면 `'.area  .box'` 처럼 스페이스가 여러개인 경우 동작하지 않습니다.

seyoon-kim replied:
테스트 후 수정하도록 하겠습니다.

dongsik-yoo commented on test/spec.js:
```diff
@@ -0,0 +1,129 @@
+    // 셀렉터가 id인 경우, 해당 id를 가지고 있는 엘리멘트 배열을 반환
+    it('if selector is id, should be return array id element', function() {
+        document.body.innerHTML = '<div class="main">main</div><div id="cont">no main</div>';
+        expect(Domutil.querySelector('#cont').length).toBe(1);
```
찾아지지 않는 Edge케이스도 테스트하면 좋을 것 같습니다. 의외로 스크립트 오류나 예외가 발생하기 쉽습니다.

seyoon-kim replied:
네 Edge에서도 확인하도록 하겠습니다

kyuwoo-choi commented on src/index.js:
```diff
@@ -0,0 +1,159 @@
+            return rootEle.getElementsByClassName(selector);
+        }
+
+        allElements = rootEle.getElementsByTagName('*');
```
별거 아닙니다만 `Ele`, `Element` 구분해서 사용하는 건가요?

seyoon-kim replied:
접두사로 붙는 것은 ele로 했으나 지금보니 통일성있게 하나로 가는 것이 좋을것 같습니다.

dongsik-yoo commented on src/index.js:
```diff
@@ -0,0 +1,159 @@
+        var founded; // founded, from 중에서 arrSeletor의 원소에 해당하는 결과를 저장하는 값
+        var from; // from, 찾아야 하는 대상이 되는 엘리멘트
+        var arrSeletor; // arrSeletor, 무엇을 찾아야 하는지 알려주는 값
+        var result = [];
```
여기도 같은 패턴이라면 초기화는 `if`문 다음에 `founded` 처럼 해야 겠네요.

seyoon-kim replied:
초기화된 변수를 상위로 올리고 선언만 된 변수는 그 아래에 작성하도록 하겠습니다.

minkyu-yi commented on test/spec.js:
```diff
@@ -0,0 +1,129 @@
+
+    // 셀렉터가 두 개인 경우, .main div의 해당하는 엘레멘트 배열을 반환
+    it('if selector is (.className tagName), should be return array .className tagName element', function() {
+        document.body.innerHTML = '<div id="cont" class="main">no main</div><div class="main"><div>.main div</div></div>';
```
innerHtml 코드가 많은데
 fixture를 만들어서 테스트하는게 더 깔끔할 것 같아요~

seyoon-kim replied:
넵 반복적인 부분은 beforeEach나 fixture를 사용하도록 하겠습니다.

seonim-ryu commented on src/index.js:
```diff
@@ -0,0 +1,159 @@
+        var rIdName = /^#/g;
+
+        if (rClassName.test(selectors)) {
+            selectors = selectors.replace(rClassName, '');
```
음.. 셀렉터들 사이 공백을 없애버리면 안되지 않을까요?
만약에 `div p` 이런식으로 태그명으로 검색을 하는 경우라면.. 이 로직에서 `divp`가 되어버리니..
문제가 있을듯 하네요~

minkyu-yi commented on src/index.js:
```diff
@@ -0,0 +1,159 @@
+        var rIdName = /^#/g;
+
+        if (rClassName.test(selectors)) {
+            selectors = selectors.replace(rClassName, '');
```
`.abc.a` 같은 셀렉터라면 `abc.a`가 만들어질 것 같은데 사전에 전처리가 되고 있는건가요~?

이름이 selectors라서 여러 셀렉터들이 조합된 string으로 생각되는데 코드 내용과는 조금 다른것 같아요

seyoon-kim replied:
.abc.a와 같이 두 개 이상의 조합된 셀렉터들은 요구 조건 사항에 포함되지 않아서 일단은 확인은 하지 않았습니다. 그러나 말씀해주신 것처럼 그러한 경우도 발생할 수 있으니 추가적으로 확인해서 테스트 해보도록 하겠습니다.

dongsik-yoo commented:
9/14 리뷰 완료입니다. 수고하셨습니다.

seonim-ryu commented on src/index.js:
```diff
@@ -0,0 +1,159 @@
+        var rIdName = /^#/g;
+
+        if (rClassName.test(selectors)) {
+            selectors = selectors.replace(rClassName, '');
```
이 로직은 공백 제거용인가요?

seyoon-kim replied:
document.getElementByClass("className")의 경우 인자로 받는 .className 문자열 앞에 '.' 를 제거하기 위해 사용하였습니다.

seonim-ryu commented:
[9/14] 리뷰 완료하였습니다. 이제부터 시작이네요ㅎ 고생하셨습니다~

kyuwoo-choi commented on src/index.js:
```diff
@@ -0,0 +1,159 @@
+        if (rClassName.test(selectors)) {
+            selectors = selectors.replace(rClassName, '');
+            result = _getElementsByClassNamePolyfill(rootEle, selectors);
+            result = Array.prototype.slice.call(result);
```
`slice.call`은 `_getElementsByClassNamePolyfill`의 마지막에서 처리해서 내보내는 것이 맞겠네요.

seyoon-kim replied:
넵 해당 함수안에 return 하기 전에 slice.call을 사용하여 배열로 바꾸었습니다.

kyuwoo-choi commented on src/index.js:
```diff
@@ -0,0 +1,159 @@
+            }
+        } else {
+            result = rootEle.getElementsByTagName(selectors);
+            result = Array.prototype.slice.call(result);
```
`[].slice.call` 짧게써도 무방합니다.

seyoon-kim replied:
넵 짧게 쓰도록 하겠습니다.

seyoon-kim commented:
바쁜시간 내주셔서 리뷰 해주셔서 감사합니다. 리뷰 해주신 부분들은 시간이 나는대로 적용하도록 하겠습니다. ^^

kyuwoo-choi commented on src/index.js:
```diff
@@ -0,0 +1,159 @@
+            if (x === 0) {
+                temp.push(from[0]);
+            } else {
+                count = 0;
```
`count`까지는 필요 없지 않을까요. boolean 으로 하고, 발견되었을때 break 해되 될 것 같습니다.

seyoon-kim replied:
넵 count대신에 indexOf 함수를 이용하여 해당 배열에 일치하는 값이 있으면 break되도록 변경하였습니다.

youjung-hong commented on src/index.js:
```diff
@@ -0,0 +1,159 @@
+                        count += 1;
+                    }
+                }
+                if (count === 0) {
```
같은 지 여부를 `count === 0`으로 판단한다면, 
같은 값이 나왔을 때 반복문에서 `break`하는 게 좋겠어요
아니면 boolean으로 `hasSameElement` 등으로 표현해도 되겠네요.

seyoon-kim replied:
넵 count대신에 indexOf 함수를 이용하여 해당 배열에 일치하는 값이 있으면 break되도록 변경하였습니다.

kyuwoo-choi commented on src/index.js:
```diff
@@ -0,0 +1,159 @@
+        for (x; x < fromLength; x += 1) {
+            if (x === 0) {
+                temp.push(from[0]);
+            } else {
```
조금 더 리팩토링해서 `indexOf`함수 하나 만들어 빼서 사용하면 더 깔끔하겠네요.

seyoon-kim replied:
for 문대신에 indexOf 함수를 이용하여 코드를 좀더 간략하게 바꾸었습니다.

kyuwoo-choi commented on src/index.js:
```diff
@@ -0,0 +1,159 @@
+
+    // selector를 만족 시키는 요소들을 담은 배열을 생성하여 반환
+    var _makeArrayMatchingToSelctor = function(selectors) {
+        var founded; // founded, from 중에서 arrSeletor의 원소에 해당하는 결과를 저장하는 값
```
http://endic.naver.com/enkrEntry.nhn?sLn=en&entryId=0cf8dae9253543448da44d7b66c9fc47&query=found

seyoon-kim replied:
founded라는 변수명은 제가 생각하기에 셀렉터로 검색하여 찾아진 결과를 담아두는 변수라고 생각하여 founded라고 사용하였습니다.

youjung-hong commented on src/index.js:
```diff
@@ -0,0 +1,159 @@
+                result = [];
+            } else {
+                result = [result];
+            }
```
이렇게 해도 되겠네요
`result = (result === null) ? [] : [result];`

seyoon-kim replied:
삼항식으로 줄여보도록 하겠습니다.

kyuwoo-choi commented on src/index.js:
```diff
@@ -0,0 +1,159 @@
+        return result;
+    };
+
+    var querySelector = function(selectors) {
```
`querySelector`는 배열이 아니라 엘리먼트를 직접 반환합니다.

seyoon-kim replied:
과제 요구조건에 배열로 반환한다고 작성하여 일단은 배열로 하였습니다. 그러나 querySelector가 본디 엘리멘트로 반환하기 때문에 추후 엘리멘트로 수정하도록 하겠습니다.

kyuwoo-choi commented on test/spec.js:
```diff
@@ -0,0 +1,129 @@
+
+describe('Domutil.querySelectorAll function should be return array ', function() {
+    // Selector API 요구조건, 배열을 반환해야 한다 테스트 케이스 작성
+    it('should be return array', function() {
```
조건이 `typeof` 이던가 `jasmine.any`를 해야 분명하겠습니다.
https://jasmine.github.io/2.0/introduction.html#section-Matching_Anything_with_<code>jasmine.any</code>

seyoon-kim replied:
jasmine.any를 이용하여 타입을 확인하도록 하였습니다.

kyuwoo-choi commented on test/spec.js:
```diff
@@ -0,0 +1,129 @@
+});
+
+describe('getElementsByClassName Pollyfil', function() {
+    var _getElementsByClassNamePolyfill = function(rootEle, selector) {
```
테스트를 하기 위해서는 부득이 Domutil에 노출해야겠네요.
코드가 여기에 있으면 안됩니다.

seyoon-kim replied:
넵 테스트할 코드는 따로 외부로 노출하여 사용하도록 하겠습니다.

kyuwoo-choi reviewed (commented):
[9/14] 수고하셨어요~

youjung-hong commented on src/index.js:
```diff
@@ -0,0 +1,159 @@
+        from = _makeArrayMatchingToSelctor(selectors);
+
+        if (from.length < 1) {
+            result = [];
```
result가 배열로 초기화되어 있으니 이부분은 필요 없겠네요

seyoon-kim replied:
넵 확인하였습니다.

youjung-hong reviewed (commented):
[9/14 코드리뷰 완료] 첫 리뷰 준비하느라 수고하셨어요~

seyoon-kim committed:
Update, 리뷰사항 반영

seyoon-kim committed:
Update, toArray, indexOf 함수 구현

하위브라우저 IE 8이하에서도 잘작동하도록 유사배열을 배열로 바꾸줘는 toArray와 배열에서 일치하는 요소가있으면 해당 인덱스를 반환해주는 indexOf 함수구현

seyoon-kim committed:
Update, 소스정리

seyoon-kim committed:
spec 정리

seyoon-kim committed:
소스정리
